### PR TITLE
Do rate limited joining

### PR DESCRIPTION
**Pull request by lukebarnard1**
I'm not sure if this is the right way to go. I've probably been a bit too extreme with the retries, so can we IRL on the exact numbers - at the moment it does 100 retries and timeouts after 10mins.

### CODE REVIEW

**Comment by lukebarnard1** [comment]
@Kegsay PTAL

File: lib/bridge/MatrixHandler.js [hunk 1]
```diff
@@ -23,6 +23,8 @@ function MatrixHandler(ircBridge) {
     };
 
     this._memberTracker = null;
+
+    this._spamming = false;
```
Comment (lukebarnard1) [inline lib/bridge/MatrixHandler.js]: Remove this

**Comment by lukebarnard1** [comment]
This refers to #199

File: lib/provisioning/Provisioner.js [hunk 2]
```diff
@@ -211,13 +211,51 @@ Provisioner.prototype._updateBridgingState = Promise.coroutine(
         }
     }
 );
+// attempts        function            args
+// retry(10, matrixClient, getStateEvent, roomId, 'm.room.power_levels');|
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: Spurious comments. Explain how `retry(attempts, obj, fn)` works instead please.

File: lib/provisioning/Provisioner.js [hunk 3]
```diff
@@ -212,12 +212,53 @@ Provisioner.prototype._updateBridgingState = Promise.coroutine(
     }
 );
 
+// Utility function for attempting to send a request to a matrix endpoint
+//  that might be rate-limited.
+//
+//  This function will try [attempts] times to apply function [fn] to [obj] by
+//  calling fn.apply(obj, args), with args being any arguments passed to retry
+//  after [fn]. If an error occurs, the same will be tried again after
+//  [retryDelayMs]. If an error err is thrown by [fn] and err.data.retry_after_ms
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: `[ ]` is generally used for arrays so this is a bit confusing. Why not use markdown-style ` ?

File: lib/provisioning/Provisioner.js [hunk 4]
```diff
@@ -212,12 +212,53 @@ Provisioner.prototype._updateBridgingState = Promise.coroutine(
...
+            return val;
+        }
+        catch (err) {
+            let msg = err.data && err.data.error ? err.data.error : err.message;
+            log.error(`Error doing rate limited action (${msg})`);
+
+            let waitTimeMs = retryDelayMS;
+
+            if (err.data && err.data.retry_after_ms && attempts > 0) {
+                attempts = attempts - 1;
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: Doesn't this mean we're now double-decrementing per interation?

**Comment by Kegsay** [comment]
Otherwise LGTM!

File: lib/provisioning/Provisioner.js (hunk 2 above)
Comment (lukebarnard1) [inline lib/provisioning/Provisioner.js]: Done.

File: lib/provisioning/Provisioner.js [hunk 5]
```diff
@@ -211,13 +211,51 @@ Provisioner.prototype._updateBridgingState = Promise.coroutine(
...
+// attempts        function            args
+// retry(10, matrixClient, getStateEvent, roomId, 'm.room.power_levels');|
+
+function retry(attempts, obj, fn) {
+    let defered = promiseutil.defer();
+
+    // Remove first two args
+    var args = Array.from(arguments).slice(3);
+
+    let tryAgain = Promise.coroutine(function*() {
```
Comment (lukebarnard1) [inline lib/provisioning/Provisioner.js]: Done.

File: lib/provisioning/Provisioner.js (hunk 3 above)
Comment (lukebarnard1) [inline lib/provisioning/Provisioner.js]: Done.

File: lib/provisioning/Provisioner.js [hunk 6]
```diff
@@ -212,12 +212,53 @@ Provisioner.prototype._updateBridgingState = Promise.coroutine(
...
+//  [retryDelayMs]. If an error err is thrown by [fn] and err.data.retry_after_ms
+//  is set, it will be added to that delay.
+//
+//  If the number of attempts is reached, an error is thrown.
+let retry = Promise.coroutine(function*(attempts, retryDelayMS, obj, fn) {
+    // Remove first 4 args
+    var args = Array.from(arguments).slice(4);
+
+    for (;attempts > 0; attempts--) {
+        log.info(`Trying to do rate limited action, ${attempts} attempts remaining.`);
```
Comment (lukebarnard1) [inline lib/provisioning/Provisioner.js]: Done.

File: lib/provisioning/Provisioner.js (hunk 4 above)
Comment (lukebarnard1) [inline lib/provisioning/Provisioner.js]: Oops. Fixed.

**Comment by Kegsay** [comment]
LGTM

**Pull request merged**